pkg/controller/migration/convert/core.go

```diff
@@ -211,210 +443,372 @@ func checkNodeHostPathVolume(spec corev1.PodSpec, name, path string) error {
+		}
 	}
+
 	nodeSel := removeOSNodeSelectors(c.node.Spec.Template.Spec.NodeSelector)
 	delete(nodeSel, "projectcalico.org/operator-node-migration")
-	if len(nodeSel) > 0 {
-		// raise error unless the only nodeSelector is the  calico-node migration nodeSelector
-		return ErrIncompatibleCluster{
-			err:       fmt.Sprintf("unsupported nodeSelector for calico-node daemonset: %v", nodeSel),
-			component: ComponentCalicoNode,
-			fix:       "remove the nodeSelector",
-		}
 
+	// Merge any remaining nodeSelectors into the installation.
+	if len(nodeSel) == 0 {
+		if hasNodeSelector(install.Spec.CalicoNodeDaemonSet) {
+			return ErrNodeSelectorOnlyOnInstall(ComponentCalicoNode)
+		}
+	} else {
+		helpers.EnsureCalicoNodeNodeSelectorNotNil(install)
+		if err := mergeNodeSelector(ComponentCalicoNode, install.Spec.CalicoNodeDaemonSet.Spec.Template.Spec.NodeSelector, nodeSel); err != nil {
+			return err
+		}
 	}
 
-	// check typha nodeSelectors
+	// Handle typha.
+	// We treat typha differently from the node and kube-controllers since there are two fields to determine typha affinity.
 	if c.typha != nil {
-		// we can migrate typha affinities provided they are a NodeAffinity for Preferred.
-		if aff := c.typha.Spec.Template.Spec.Affinity; aff != nil {
-			if aff.PodAffinity != nil || aff.PodAntiAffinity != nil {
-				return ErrIncompatibleCluster{
-					err:       "pod affinity and antiAffinity not supported for typha deployment",
-					component: ComponentTypha,
-					fix:       "remove the affinity",
+		installAffinity := getTyphaAffinity(install)
+		if c.typha.Spec.Template.Spec.Affinity != nil {
+			if installAffinity != nil {
+				// Affinity is set on the component and the installation, verify that they match.
+				if !reflect.DeepEqual(c.typha.Spec.Template.Spec.Affinity, installAffinity) {
+					return ErrIncompatibleAffinity(ComponentTypha)
 				}
 			}
-			if aff.NodeAffinity != nil {
-				if aff.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
-					return ErrIncompatibleCluster{
-						err:       "nodeAffinity 'RequiredDuringSchedulingIgnoredDuringExecution' not supported on Typha.",
-						component: ComponentTypha,
-						fix:       "remove the affinity",
-					}
-				}
-				install.Spec.TyphaAffinity = &operatorv1.TyphaAffinity{
-					NodeAffinity: &operatorv1.NodeAffinity{
-						PreferredDuringSchedulingIgnoredDuringExecution: aff.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution,
-					},
-				}
+			// Ensure that the affinity on the installation uses the new field and clears the deprecated TyphaAffinity.
+			helpers.EnsureTyphaPodSpecNotNil(install)
+			install.Spec.TyphaDeployment.Spec.Template.Spec.Affinity = c.typha.Spec.Template.Spec.Affinity
+			install.Spec.TyphaAffinity = nil
+		} else {
+			// Affinity is not set on the component, check if it's set on the installation.
+			if installAffinity != nil {
+				return ErrIncompatibleAffinity(ComponentTypha)
```

Comment:
I think the user adding affinity to the install would be fine, why should we error?